### PR TITLE
ODC-7179 : Allow users to see which pods are receiving traffic

### PR DESCRIPTION
**Pull request by lokanandaprabhu**
JIRA story:
https://issues.redhat.com/browse/ODC-7179

Description:
Added traffic details for pods in below places in the console UI,
1) In the Resources tab of the topology side panel when showing pods. Along with icon tooltip is also added to indicate whether pod is receiving traffic or not.
2) In the Pods tab of related resources ( Service, Deployment, Deployment Config)
3) In the Pods list view in the context of a namespace, add a traffic column (talk to UX/Sam about how we can add traffic info in the list)
4) In Pods list view / All Namespaces, do not add the traffic column
5) If a Pod is receiving traffic, show that on the Pod details page
 

Demo:

https://user-images.githubusercontent.com/102503482/202167364-98b4c8e9-d5ed-43f4-b599-55e04c3933f7.mov

How to test:
Create the deployment or DC or service and check the traffic details for pods in below places in console, 
1) In the Resources tab of the topology side panel when showing pods. Along with icon tooltip is also added to indicate whether pod is receiving traffic or not.
2) In the Pods tab of related resources ( Service, Deployment, Deployment Config)
3) In the Pods list view in the context of a namespace, add a traffic column (talk to UX/Sam about how we can add traffic info in the list)
4) In Pods list view / All Namespaces, do not add the traffic column
5) If a Pod is receiving traffic, show that on the Pod details page

### CODE REVIEW

**Comment by lokanandaprabhu** [comment]
Hi @serenamarie125 / @debsmita1 ,

Created the PR for the issue https://issues.redhat.com/browse/ODC-7179. I have added demo video as well. PTAL. Thanks.

File: frontend/public/components/utils/get-pod-traffic.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { useK8sWatchResource } from '@console/internal/components/utils/k8s-watch-hook';
+import { EndpointSliceKind } from '@console/internal/module/k8s';
+import { EndPointSliceModel } from '../../../packages/console-app/src/models';
```
Comment (invincibleJai) [inline frontend/public/components/utils/get-pod-traffic.tsx:3]: should this be imported from `@console/shared` over relative import?

File: frontend/public/components/utils/get-pod-traffic.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import { ConnectedIcon, DisconnectedIcon } from '@patternfly/react-icons';
+import * as React from 'react';
+import { Tooltip } from '@patternfly/react-core';
+import { useTranslation } from 'react-i18next';
```
Comment (invincibleJai) [inline frontend/public/components/utils/get-pod-traffic.tsx:7]: sort the imports, these should go first

File: frontend/public/components/utils/get-pod-traffic.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  namespace: string;
+  tooltipFlag?: boolean;
+};
+
+export const GetPodTraffic: React.FC<GetPodTrafficProp> = ({ podName, namespace, tooltipFlag }) => {
+  const { t } = useTranslation();
+  const [data, loaded, loadError] = useK8sWatchResource<EndpointSliceKind[]>({
+    groupVersionKind: {
+      kind: EndPointSliceModel.kind,
+      version: 'v1',
```
Comment (invincibleJai) [inline frontend/public/components/utils/get-pod-traffic.tsx:20]:
```suggestion
      version: EndPointSliceModel.apiVersion,
```

**Comment by lokanandaprabhu** [comment]
> on load icons show not receiving initially and then it showed receiving traffic, is this expected?
> 
> Thanks @lokanandaprabhu , have some observations
> 
> 1. on load shall we show `-` or loading indicator rather than defaulting to no traffic
> 
> receivingTraffic [ receivingTraffic ](https://user-images.githubusercontent.com/5129024/202459383-d575ae6e-93d3-4b4c-8a71-01c30382718a.gif) [ ](https://user-images.githubusercontent.com/5129024/202459383-d575ae6e-93d3-4b4c-8a71-01c30382718a.gif)
> 
> 2. why do we don't show receiving traffic for all projects even though it's a selected column, is this based on direction from epic  (as this could be confusing to the user)
> 
> image


Point 2 based on below AC. 
> In Pods list view / All Namespaces, do not add the traffic column

I will think about point 1, whether to add '-' while loading or add loading indicator.

File: frontend/public/components/pod.tsx [hunk 4]
```diff
@@ -114,6 +115,8 @@ import VirtualizedTable, { TableData } from './factory/Table/VirtualizedTable';
 import { sortResourceByValue } from './factory/Table/sort';
 import { useActiveColumns } from './factory/Table/active-columns-hook';
 import { PodDisruptionBudgetField } from '@console/app/src/components/pdb/PodDisruptionBudgetField';
+import { GetPodTraffic } from './utils/get-pod-traffic';
+import { useActiveNamespace } from '@console/shared/src/hooks/useActiveNamespace';
```
Comment (invincibleJai) [inline frontend/public/components/pod.tsx:119]: nit: sort import

File: frontend/public/components/utils/get-pod-traffic.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    groupVersionKind: {
+      kind: EndPointSliceModel.kind,
+      version: EndPointSliceModel.apiVersion,
+    },
+    isList: true,
+    namespaced: true,
+    namespace,
+  });
+
+  const allEndpoints = data.reduce((prev, next) => prev.concat(next?.endpoints), []);
```
Comment (invincibleJai) [inline frontend/public/components/utils/get-pod-traffic.tsx:27]: wold be better to check if loaded, no load error, and if data is present before processing?

File: frontend/public/components/utils/get-pod-traffic.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  const receivingTraffic = allEndpoints.find((obj) => obj?.targetRef?.name === podName);
+
+  if (tooltipFlag) {
```
Comment (invincibleJai) [inline frontend/public/components/utils/get-pod-traffic.tsx:30]: do we need this i.e when we need tooltips and when we don't? even in the list, we can show the tooltip right?

File: frontend/public/module/k8s/types.ts [hunk 7]
```diff
@@ -1200,3 +1200,16 @@ export type ReplicationControllerKind = {
...
+    {
+      targetRef?: {
+        kind?: string;
+        name?: string;
+        namespace?: string;
+        uid?: string;
+      };
+    },
+  ];
+} & K8sResourceCommon;
```
Comment (invincibleJai) [inline frontend/public/module/k8s/types.ts:1215]: There could be multiple items in `endpoints` list right?

`endpoints is a list of unique endpoints in this slice. Each slice may include a maximum of 1000 endpoints.`

```suggestion
type EndpointSlice = {
        kind?: string;
        name?: string;
        namespace?: string;
        uid?: string;
} ;

export type EndpointSliceKind = {
  endpoints?: EndpointSlice[];
} & K8sResourceCommon;
```

File: frontend/public/components/utils/get-pod-traffic.tsx (hunk 6 above)
Comment (lokanandaprabhu) [reply frontend/public/components/utils/get-pod-traffic.tsx:30]: Tooltip is required only for topology side bar in pod details section. So not added tooltip for list pages. This is based on the conversation - https://coreos.slack.com/archives/CHGU4P8UU/p1668540290287939?thread_ts=1668424861.378819&cid=CHGU4P8UU

File: frontend/public/components/overview/pods-overview.tsx [hunk 8]
```diff
@@ -179,7 +179,10 @@ export const PodsOverviewContent: React.SFC<PodsOverviewContentProps> = ({
       variant="info"
       title={t('public~Waiting for the build')}
       actionLinks={
-        <AlertActionLink onClick={() => setShowWaitingPods(!showWaitingPods)}>
+        <AlertActionLink
+          onClick={() => setShowWaitingPods(!showWaitingPods)}
+          data-test="waiting-pods"
+        >
```
Comment (invincibleJai) [inline frontend/public/components/overview/pods-overview.tsx:185]: not sure if this is consumed as I see it 

https://github.com/openshift/console/pull/12272/files#diff-982e6cfb636ca8ca63999882c646715710f983c0663a05dfd1e46c17e7714075R187-R190

File: frontend/public/components/utils/get-pod-traffic.tsx [hunk 9]
```diff
@@ -0,0 +1,55 @@
+import * as React from 'react';
+import { useTranslation } from 'react-i18next';
+import { Tooltip } from '@patternfly/react-core';
+import { useK8sWatchResource } from '@console/internal/components/utils/k8s-watch-hook';
+import { EndpointSliceKind } from '@console/internal/module/k8s';
```
Comment (invincibleJai) [inline frontend/public/components/utils/get-pod-traffic.tsx:5]: this file is in the public package so can do a relative import?

**Comment by invincibleJai** [comment]
/cc @sanketpathak 

needs your help with review of e2e

File: frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature [hunk 10]
```diff
@@ -0,0 +1,26 @@
+@Pod-Traffic-Status
+Feature: Traffic Status details for pods
+              As a administrator, I want to see whether a pod is receiving traffic or not.
+
+        Background:
+            Given user is at developer perspective
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature:6]:
```suggestion
            Given user is at administrator perspective
```

File: frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+              And user has created or selected namespace "aut-pods"
+              And user is at Add page
+ 
+        @regression 
+        Scenario: Checking traffic status for pods in a project: P-01-TC01
+            Given user has created workload "nodejs-ex-git" with resource type "deployment"
+             When user is at admin perspective
+              And user navigates to pods tab
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature:14]:
```suggestion
             When user navigates to pods tab
```

File: frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature [hunk 12]
```diff
@@ -0,0 +1,26 @@
+@Pod-Traffic-Status
+Feature: Traffic Status details for pods
+              As a administrator, I want to see whether a pod is receiving traffic or not.
+
+        Background:
+            Given user is at developer perspective
+              And user has created or selected namespace "aut-pods"
+              And user is at Add page
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature:8]:
```suggestion
              And user navigates to pods tab
              And user has created or selected namespace "aut-pods"
```

File: frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+             When user is at admin perspective
+              And user navigates to pods tab
+              And user selects Receiving Traffic column to show in table
+             Then user is able to see Receiving Traffic column in the list
+
+        @regression 
+        Scenario: Checking traffic status for pods for all projects: P-01-TC02
+            Given user has created workload "nodejs-ex-git" with resource type "deployment"
+              And user is at admin perspective
+              And user navigates to pods tab
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature:22]:
```suggestion
```

File: frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+              As a administrator, I want to see whether a pod is receiving traffic or not.
+
+        Background:
+            Given user is at developer perspective
+              And user has created or selected namespace "aut-pods"
+              And user is at Add page
+ 
+        @regression 
+        Scenario: Checking traffic status for pods in a project: P-01-TC01
+            Given user has created workload "nodejs-ex-git" with resource type "deployment"
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature:12]:
```suggestion
```

File: frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature (hunk 13 above)
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/pod-list/pod-list-page.feature:22]:
```suggestion
```

**Comment by lokanandaprabhu** [comment]
Hi @sanketpathak ,

Updated e2e tests. PTAL. Thanks.

**Comment by sanketpathak** [comment]
Verified on cluster created on pr using cluster-bot
Screenshot from 2022-12-02 20-30-16
Screenshot from 2022-12-02 20-11-44
Screenshot from 2022-12-02 19-56-20
Screenshot from 2022-12-02 19-54-46

/label qe-approved

**Comment by lokanandaprabhu** [comment]
Hi @debsmita1 ,

Based on @jerolimov suggestion I changed the component name from GetPodTraffic to PodTraffic. PTAL.

**Review by jerolimov — APPROVED** [review]
Thanks for moving the component. As nothing other has changed I restore the lgtm from @debsmita1 without taking a deep look. :smirk:

/lgtm
/approve

**Comment by lokanandaprabhu** [comment]
Updated i18n.

**Pull request merged**